hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -867,18 +867,20 @@ object HoodieSparkSqlWriter {
           hoodieRecord
         }).toJavaRDD()
       case HoodieRecord.HoodieRecordType.SPARK =>
+        log.info("Use spark record")
         // ut will use AvroKeyGenerator, so we need to cast it in spark record
         val sparkKeyGenerator = keyGenerator.asInstanceOf[SparkKeyGeneratorInterface]
+        val schemaWithMetaField = HoodieAvroUtils.addMetadataFields(schema, config.allowOperationMetadataField)
         val structType = HoodieInternalRowUtils.getCachedSchema(schema)
-        val structTypeBC = SparkContext.getOrCreate().broadcast(structType)
-        HoodieInternalRowUtils.addCompressedSchema(structType)
+        val structTypeWithMetaField = HoodieInternalRowUtils.getCachedSchema(schemaWithMetaField)
+        val structTypeBC = sparkContext.broadcast(structType)
+        HoodieInternalRowUtils.broadcastCompressedSchema(List(structType, structTypeWithMetaField), sparkContext)
```

Comment:
Why do we need this?

Reply:
The broadcast ensures that every executor has the same fingerprint->schema cache. We can't make every executor initialize fingerprint->schema locally. 

For example, if we initialize fingerprint->schema in sparksqlwriter df, it is possible that **the number of dataframe partitions is less than the number of executors**.

```scala
        HoodieInternalRowUtils.addCompressedSchema(structType)
        df.queryExecution.toRdd.map(row => {
          val internalRow = row.copy()
          val (processedRow, writeSchema) = getSparkProcessedRecord(partitionCols, internalRow, dropPartitionColumns, structTypeBC.value)
          val recordKey = sparkKeyGenerator.getRecordKey(internalRow, structTypeBC.value)
          val partitionPath = sparkKeyGenerator.getPartitionPath(internalRow, structTypeBC.value)
          val key = new HoodieKey(recordKey.toString, partitionPath.toString)
          // See this We can't guarantee that this is executed in all executors
          HoodieInternalRowUtils.addCompressedSchema(structType)

          new HoodieSparkRecord(key, processedRow, writeSchema)
        }).toJavaRDD().asInstanceOf[JavaRDD[HoodieRecord[_]]]
```
As shown in the figure below, executor3 do not has the fingerPrint cache and can not deserialize the fingerPrint to schema.
![未命名文件](https://user-images.githubusercontent.com/25837418/192523203-d32dd632-f5b9-4b10-9f5f-ad733ce787d1.png)

Reply:
Sorry, i misread your original intent with this change. Unfortunately, this is not going to work: for Broadcast to work exactly the same object has to be passed in (w/in captured Closure) from Driver to Executor (Broadcasts are identified by unique `bid`). In the current setup you're just overwriting static variables w/in `HoodieInternalRowUtils` creating new Broadcasts which aren't propagated to Executors (the reason it works in tests is b/c we run phony Spark cluster where both Driver and Executor share memory, ie they access the same loaded `HoodieInternalRowUtils` and this problem is not apparent)

Reply:
We need a simpler solution here:

 - We should remove the schema from `HoodieSparkRecord`s. At this point we ultimately are familiar with all the limitations of the approach which Spark uses to reduce overhead of serializing the schema along w/ every record -- it requires us to provide all the schemas upfront to be distributed to all Executors (via `SparkConf` which is read-only after `SparkContext` is created).

 - We will have to pass record's schema externally: we can approach it the same way it's currently done for `ExpressionPayload` in [this PR](https://github.com/apache/hudi/pull/6358#pullrequestreview-1122556175)

Reply:
I don't agree with that Broadcast does not work. I have tested this on yarn and it worked. The process of broadcast: the driver will put the blockcast variable into the blockmanager. And executors will be obtained locally according to the block id, and if it is not available locally, it will be pulled from the remote driver.

If the Broadcast works, we do not need to pass record's schema externally right? Passing record's schema externally requires user of HoodieRecord to pass the right schema.